File: public/api/component-images.php
<?php

require dirname(__DIR__, 2) . '/vendor/autoload.php';

use App\Database;
use App\ComponentRepository;
use App\ImageRepository;
use App\TagRepository;
use App\ComponentLoader;

header('Content-Type: application/json');

$method = $_SERVER['REQUEST_METHOD'];
$db     = Database::get();
$repo   = new ComponentRepository($db);

if ($method === 'GET') {
    $componentId = $_GET['component_id'] ?? '';
    if ($componentId === '') {
        http_response_code(400);
        echo json_encode(['error' => 'Missing component_id']);
        exit;
    }

    $rows    = $repo->forComponent($componentId);
    $tagRepo = new TagRepository($db);
    $imageIds = array_column($rows, 'id');
    $tagMap   = $tagRepo->tagsForImages($imageIds);

    $images = array_map(fn($r) => [
        'assignment_id' => (int) $r['assignment_id'],
        'is_active'     => (bool) $r['is_active'],
        'assigned_at'   => $r['assigned_at'],
        'id'            => $r['id'],
        'original_name' => $r['original_name'],
        'name'          => $r['name'] ?: $r['original_name'],
        'filename'      => $r['filename'],
        'width'         => (int) $r['width'],
        'height'        => (int) $r['height'],
        'file_size'     => (int) $r['file_size'],
        'tags'          => $tagMap[$r['id']] ?? [],
        'thumb_url'     => 'image-serve.php?file=' . urlencode($r['filename']) . '&w=300&h=200&fit=crop',
    ], $rows);

    echo json_encode(['images' => $images]);
    exit;
}

if ($method === 'POST') {
    $body        = json_decode(file_get_contents('php://input'), true);
    $componentId = $body['component_id'] ?? '';
    $imageId     = $body['image_id']     ?? '';

    if ($componentId === '' || $imageId === '') {
        http_response_code(400);
        echo json_encode(['error' => 'Missing component_id or image_id']);
        exit;
    }

    if (ComponentLoader::findComponent($componentId) === null) {
        http_response_code(404);
        echo json_encode(['error' => 'Component not found']);
        exit;
    }

    $imgRepo = new ImageRepository($db);
    if ($imgRepo->findById($imageId) === null) {
        http_response_code(404);
        echo json_encode(['error' => 'Image not found']);
        exit;
    }

    $repo->assign($componentId, $imageId);
    echo json_encode(['success' => true]);
    exit;
}

if ($method === 'PATCH') {
    $body         = json_decode(file_get_contents('php://input'), true);
    $assignmentId = (int) ($body['assignment_id'] ?? 0);

    if ($assignmentId === 0) {
        http_response_code(400);
        echo json_encode(['error' => 'Missing assignment_id']);
        exit;
    }

    $repo->setActive($assignmentId);
    echo json_encode(['success' => true]);
    exit;
}

if ($method === 'DELETE') {
    $assignmentId = (int) ($_GET['id'] ?? 0);

    if ($assignmentId === 0) {
        http_response_code(400);
        echo json_encode(['error' => 'Missing id']);
        exit;
    }

    $repo->removeAssignment($assignmentId);
    echo json_encode(['success' => true]);
    exit;
}

http_response_code(405);
echo json_encode(['error' => 'Method not allowed']);

File: internal/handler/component_images.go
package handler

import (
	"imagemanager/internal/config"
	"imagemanager/internal/repo"
	"net/http"
	"net/url"
	"strconv"
)

func (a *App) ComponentImages(w http.ResponseWriter, r *http.Request) {
	compRepo := repo.NewComponentRepo(a.DB)

	switch r.Method {
	case http.MethodGet:
		componentID := r.URL.Query().Get("component_id")
		if componentID == "" {
			jsonErr(w, 400, "Missing component_id")
			return
		}

		rows, err := compRepo.ForComponent(componentID)
		if err != nil {
			jsonErr(w, 500, "DB error")
			return
		}

		tagRepo := repo.NewTagRepo(a.DB)
		var imageIDs []string
		for _, r := range rows {
			imageIDs = append(imageIDs, r.ImageID)
		}
		tagMap, _ := tagRepo.ForImages(imageIDs)

		images := make([]map[string]any, len(rows))
		for i, ci := range rows {
			images[i] = map[string]any{
				"assignment_id": ci.AssignmentID,
				"is_active":     ci.IsActive,
				"assigned_at":   ci.AssignedAt,
				"id":            ci.ImageID,
				"original_name": ci.OriginalName,
				"name":          ci.Name,
				"filename":      ci.Filename,
				"width":         ci.Width,
				"height":        ci.Height,
				"file_size":     ci.FileSize,
				"tags":          tagMap[ci.ImageID],
				"thumb_url":     "image-serve?file=" + url.QueryEscape(ci.Filename) + "&w=300&h=200&fit=crop",
			}
			if images[i]["tags"] == nil {
				images[i]["tags"] = []interface{}{}
			}
		}
		jsonOK(w, map[string]any{"images": images})

	case http.MethodPost:
		var body struct {
			ComponentID string `json:"component_id"`
			ImageID     string `json:"image_id"`
		}
		if err := readJSON(r, &body); err != nil || body.ComponentID == "" || body.ImageID == "" {
			jsonErr(w, 400, "Missing component_id or image_id")
			return
		}

		components, _ := config.LoadComponents(a.AppRoot)
		if config.FindComponent(components, body.ComponentID) == nil {
			jsonErr(w, 404, "Component not found")
			return
		}

		imgRepo := repo.NewImageRepo(a.DB)
		if img, _ := imgRepo.FindByID(body.ImageID); img == nil {
			jsonErr(w, 404, "Image not found")
			return
		}

		if err := compRepo.Assign(body.ComponentID, body.ImageID); err != nil {
			jsonErr(w, 500, "Failed to assign image")
			return
		}
		jsonOK(w, map[string]bool{"success": true})

	case http.MethodPatch:
		var body struct {
			AssignmentID int64 `json:"assignment_id"`
		}
		if err := readJSON(r, &body); err != nil || body.AssignmentID == 0 {
			jsonErr(w, 400, "Missing assignment_id")
			return
		}
		if err := compRepo.SetActive(body.AssignmentID); err != nil {
			jsonErr(w, 500, "DB error")
			return
		}
		jsonOK(w, map[string]bool{"success": true})

	case http.MethodDelete:
		idStr := r.URL.Query().Get("id")
		id, err := strconv.ParseInt(idStr, 10, 64)
		if err != nil || id == 0 {
			jsonErr(w, 400, "Missing id")
			return
		}
		if err := compRepo.RemoveAssignment(id); err != nil {
			jsonErr(w, 500, "DB error")
			return
		}
		jsonOK(w, map[string]bool{"success": true})

	default:
		jsonErr(w, 405, "Method not allowed")
	}
}
